lib/FrontendTool/FrontendTool.cpp

```diff
@@ -453,9 +482,44 @@ void ABIDependencyEvaluator::reexposeImportedABI(
     addToABIExportMap(module, reexportedModule);
 }
 
+bool ABIDependencyEvaluator::isOverlayOfClangModule(ModuleDecl *swiftModule) {
+  assert(!swiftModule->isNonSwiftModule());
+
+  auto cacheEntry = isOverlayCache.find(swiftModule);
+  if (cacheEntry != isOverlayCache.end())
+    return cacheEntry->second;
+
+  llvm::SmallPtrSet<ModuleDecl *, 8> importList;
+  ::getImmediateImports(swiftModule, importList,
+                        {ModuleDecl::ImportFilterKind::Public});
+  bool isOverlay =
+      llvm::any_of(importList, [&](ModuleDecl *importedModule) -> bool {
+        return isClangOverlayOf(swiftModule, importedModule);
+      });
+  isOverlayCache[swiftModule] = isOverlay;
```

Comment:
How expensive is this? Naïvely, it looks like we're caching an O(n) computation; is it hot enough to be worth the complexity of the cache?

Reply:
The complexity is incurred on a per-import basis. This means that if I have `N_o` overlays, each such overlay has `N_i` import statements and each overlay is imported by `N_c` Clang modules, then we will have a `N_c * N_i * N_o` computation instead of a `N_i * N_o` computation.

Reply:
I measured this with an example module which imports many modules (and does just that); the cache saves about 0.2 ms ~ 0.4 ms per trace. Which doesn't seem like much at all, so I'm going to get rid of it.